src/shared/scanner.rs

```diff
@@ -0,0 +1,403 @@
+                        self.make_token(TokenType::BangEqual)
+                    } else {
+                        self.make_token(TokenType::Bang)
+                    }
+                }
+                '=' => {
+                    if self.advance_if_match('=') {
+                        self.make_token(TokenType::EqualEqual)
+                    } else {
+                        self.make_token(TokenType::Equal)
+                    }
+                }
+                '<' => {
+                    if self.advance_if_match('=') {
+                        self.make_token(TokenType::LessEqual)
+                    } else {
+                        self.make_token(TokenType::Less)
+                    }
+                }
+                '>' => {
+                    if self.advance_if_match('=') {
+                        self.make_token(TokenType::GreaterEqual)
+                    } else {
+                        self.make_token(TokenType::Greater)
+                    }
+                }
+                '/' => {
+                    if self.advance_if_match('/') {
+                        self.advance_while(|c| c != '\n');
+                        self.make_token(TokenType::Comment(self.lexeme()))
+                    } else {
+                        self.make_token(TokenType::Slash)
+                    }
+                }
+                '"' => {
+                    while let Some((_, c)) = self.advance() {
+                        if c == '"' {
+                            return self.make_token(TokenType::String(
+                                // Adjusting the bounds manually here to strip the quotes off is safe,
+                                // because we know that the lexeme is bounded by ASCII quote characters.
+                                self.source[self.lexeme_start + 1..self.current_offset].into(),
+                            ));
+                        }
+                    }
+                    Err(ScannerError::UnterminatedString { line: self.line })
+                }
+                '0'..='9' => {
+                    self.advance_while(|c| c.is_ascii_digit());
+
+                    if self.peek().is_some_and(|c| c == '.')
+                        && self.peek_peek().is_some_and(|c| c.is_ascii_digit())
+                    {
+                        self.advance(); // consume the .
+                        self.advance_while(|c| c.is_ascii_digit());
+                    }
+
+                    if let Ok(number) = self.lexeme().parse() {
+                        self.make_token(TokenType::Number(number))
+                    } else {
+                        Err(ScannerError::InvalidNumber {
+                            line: self.line,
+                            number: self.lexeme().into(),
+                        })
+                    }
+                }
+                'a'..='z' | 'A'..='Z' | '_' => {
+                    self.advance_while(|c| c.is_alphanumeric() || c == '_');
```

Comment:
Is `is_alphanumeric` broader than ascii-only ?

Reply:
Seems like yes, but doesn't include emoji? https://doc.rust-lang.org/std/primitive.char.html#method.is_alphabetic

I'm thinking of using my work time to do full unicode identifiers like you did!